src/main/java/org/prebid/server/bidder/undertone/UndertoneBidder.java

```diff
@@ -0,0 +1,217 @@
+            new TypeReference<>() {
+            };
+
+    private static final int ADAPTER_ID = 3;
+    private static final String VERSION = "1.0.0";
+    private static final ObjectNode BIDDER_PARAMS = makeBidderParams();
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public UndertoneBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public final Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        if (!ObjectUtils.anyNotNull(bidRequest.getSite(), bidRequest.getApp())) {
+            return Result.withError(BidderError.badInput("invalid bidRequest: no App/Site objects"));
+        }
+
+        final Map<String, ExtImpUndertone> undertoneExtImpMap = getUndertoneExtImpMap(bidRequest);
+
+        final Integer publisherId = getPublisherId(undertoneExtImpMap);
+        if (publisherId == null) {
+            return Result.withError(BidderError.badInput("invalid bidRequest: no publisher-id"));
+        }
+
+        final List<Imp> imps = makeImps(bidRequest, undertoneExtImpMap);
+        if (imps.isEmpty()) {
+            return Result.withError(BidderError.badInput("invalid bidRequest: no valid imps"));
+        }
+
+        final BidRequest undertoneBidRequest = makeBidRequest(bidRequest, publisherId, imps);
+
+        final HttpRequest<BidRequest> httpRequest = HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .body(mapper.encodeToBytes(undertoneBidRequest))
+                .payload(bidRequest)
+                .build();
+
+        return Result.withValue(httpRequest);
+    }
+
+    @Override
+    public final Result<List<BidderBid>> makeBids(BidderCall<BidRequest> httpCall, BidRequest bidRequest) {
+        try {
+            final BidResponse bidResponse = mapper.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.withValues(extractBids(httpCall.getRequest().getPayload(), bidResponse));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private BidRequest makeBidRequest(BidRequest bidRequest, int publisherId, List<Imp> imps) {
+        final Publisher publisher = makePublisher(bidRequest, publisherId);
+
+        final ExtRequestPrebid extRequestPrebid = ExtRequestPrebid.builder()
+                .bidderparams(BIDDER_PARAMS)
+                .build();
+
+        final BidRequest.BidRequestBuilder bidRequestBuilder = bidRequest.toBuilder()
+                .imp(imps)
+                .ext(ExtRequest.of(extRequestPrebid));
+
+        if (bidRequest.getSite() != null) {
```

Comment:
Pls extract `bidRequest.getSite()` to separate variable, as you use it more then once